src/main/java/uk/gov/justice/digital/zone/StructuredZone.java

```diff
@@ -78,96 +68,60 @@ public Dataset<Row> process(
-                dataFrame,
-                sourceReference.getSchema(),
-                sourceReference.getSource(),
-                sourceReference.getTable()
-        );
+        val validatedDataFrame = validateJsonData(spark, dataFrame, sourceReference.getSchema(), source, table);
 
-        handleInvalidRecords(
-                spark,
-                validatedDataFrame,
-                sourceReference.getSource(),
-                sourceReference.getTable(),
-                validationFailedViolationPath
-        );
+        handleInvalidRecords(spark, validatedDataFrame, source, table, validationFailedViolationPath);
 
-        return handleValidRecords(spark, validatedDataFrame, tablePath, sourceReference.getPrimaryKey(), isCDC);
+        return handleValidRecords(spark, validatedDataFrame, tablePath, sourceReference.getPrimaryKey());
     }
 
-    private Dataset<Row> validateJsonData(SparkSession spark,
-                                          Dataset<Row> dataFrame,
-                                          StructType schema,
-                                          String source,
-                                          String table) {
-
-        logger.info("Validating data against schema: {}/{}", source, table);
-
-        val jsonValidator = JsonValidator.createAndRegister(schema, spark, source, table);
-
-        return dataFrame
-                .select(col(DATA), col(METADATA), col(OPERATION))
-                .withColumn(PARSED_DATA, from_json(col(DATA), schema, jsonOptions))
-                .withColumn(VALID, jsonValidator.apply(col(DATA), to_json(col(PARSED_DATA), jsonOptions)));
-    }
+    private Dataset<Row> handleNoSchemaFound(
+            SparkSession spark,
+            Dataset<Row> dataFrame,
+            String source,
+            String table
+    ) throws DataStorageException {
+        logger.error("Structured Zone Violation - No schema found for {}/{} - writing {} records",
+                source,
+                table,
+                dataFrame.count()
+        );
 
-    private Dataset<Row> handleValidRecords(SparkSession spark,
-                                            Dataset<Row> dataFrame,
-                                            String destinationPath,
-                                            SourceReference.PrimaryKey primaryKey,
-                                            Boolean isCDC) throws DataStorageException {
-        val validRecords = dataFrame
-                .filter(col(VALID).equalTo(true))
-                .select(PARSED_DATA + ".*", OPERATION);
+        val missingSchemaRecords = dataFrame
+                .select(col(DATA), col(METADATA))
+                .withColumn(ERROR, lit(String.format("Schema does not exist for %s/%s", source, table)));
 
-        val validRecordsCount = validRecords.count();
+        writeInvalidRecords(spark, storage, createValidatedPath(violationsPath, source, table), missingSchemaRecords);
 
-        if (validRecordsCount > 0) {
```

Comment:
don't call the spark action here. instead check validRecords is empty or not.

validRecords.isEmpty()

Reply:
We use the count in the log messages at many points in the code.
I agree with you. For performance reasons it would be preferred to not force spark to perform an action unless necessary. I intend to revisit the points where this is done when we get to a performance optimisation stage.

Reply:
Yeah, this is my bad. Would be good to retain the counts using a more efficient method since we can derive throughput metrics from these values.